src/main/java/hudson/plugins/deploy/CargoContainerAdapter.java

```diff
@@ -70,15 +72,36 @@ protected void deploy(DeployerFactory deployerFactory, final BuildListener liste
             if (!StringUtils.isEmpty(contextPath)) {
                 war.setContext(contextPath);
             }
-            deployer.redeploy(war);
+            deployWithRetries(deployer, listener, war, attempts);
         } else if ("EAR".equalsIgnoreCase(extension)) {
             EAR ear = createEAR(f);
-            deployer.redeploy(ear);
+            deployWithRetries(deployer, listener, ear, attempts);
         } else {
             throw new RuntimeException("Extension File Error.");
         }
     }
 
+    protected void deployWithRetries(Deployer deployer, BuildListener listener, Deployable war, int attempts) {
+        for (int tries = 1; tries <= attempts; tries++) {
+            try {
+                deployer.redeploy(war);
+                return;
+            } catch (ContainerException ce) {
+                listener.getLogger().println("Deploy Problem [attempt " + tries + " of " + attempts + "] - " + ce.getMessage());
+                if (tries >= attempts) {
+                    throw ce;
+                } else {
+                    try {
+                        // Wait a little bit, the tomcat could be restarting, then try it again.
+                        Thread.sleep(15 * 1000);
```

Comment:
Are you talking about the wait time between attempts? The number of attempts is already configurable. I thought about making the wait time configurable as well, it felt like it would be of dubious benefit. I agree 15 seconds seems a little arbitrary but for example, restarting Tomcat instances that already have web-apps deployed, it seemed like 15 seconds could be a good lower end for how long for Tomcat to complete restarting and be ready for another deployment attempt.